src/Rule/Validate/UploadedFile.php

```diff
@@ -0,0 +1,241 @@
+     *  - fileMedia     : array or strign of acceptable file media types
+     *  - sizeMax       : bytes or human readable string
+     *  - sizeMin       : bytes or human readable string
+     *
+     * @param object $subject The subject to be filtered.
+     *
+     * @param string $field The subject field name.
+     *
+     * @return bool True if valid, false if not.
+     *
+     */
+    public function __invoke($subject, $field, array $options)
+    {
+        $value = $subject->$field;
+
+        if (! $this->isRequired($options)
+            && ! $this->fileWasUploaded($value)
+        ) {
+            return true;
+        }
+
+        foreach ($this->rules as $rule) {
+            if (array_key_exists($rule, $options)) {
+                if (! $this->$rule($value, $options[$rule])) {
+                    return false;
+                }
+            }
+        }
+        // looks like we're ok!
+        return true;
+    }
+
+    /**
+     * If the file required to be uploaded?
+     *
+     * @param array $options the options passed tot the filter
+     *
+     * @return bool
+     *
+     * @access protected
+     */
+    protected function isRequired(array $options)
+    {
+        return (
+            isset($options[self::REQUIRED])
+            && $options[self::REQUIRED] == true
+        );
+    }
+
+    /**
+     * Is value an instance of a PSR7 UploadedFileInterface?
+     *
+     * @param mixed $value value being validated
+     *
+     * @return bool
+     *
+     * @access protected
+     */
+    protected function isUploadedFile($value)
+    {
+        return ($value instanceof UploadedFileInterface);
+    }
+
+    /**
+     * Was a file uploaded ?
+     *
+     * @param mixed $value value being validated
+     *
+     * @return bool
+     *
+     * @access protected
+     */
+    protected function fileWasUploaded($value)
+    {
+        if (! $this->isUploadedFile($value)) {
+            return false;
+        }
+
+        if ($value->getError() !==  UPLOAD_ERR_OK) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Required Rule Test
+     *
+     * @param mixed $value  value being validated
+     * @param bool  $option if the file is required
+     *
+     * @return bool
+     *
+     * @access protected
+     */
+    protected function required($value, $option)
+    {
+        if ($option === false) {
+            return true;
+        }
+
+        return $this->fileWasUploaded($value);
+    }
+
+    /**
+     * FileExtension Rule Test
+     *
+     * @param mixed        $value value being validated
+     * @param string|array $exts  array or string of valid extensions
+     *
+     * @return bool
+     *
+     * @access protected
+     */
+    protected function fileExtension($value, $exts)
+    {
+        $filename = $value->getClientFilename();
+        $ext = strtolower(pathinfo($filename, PATHINFO_EXTENSION));
```

Comment:
Rather than validating based on the pathinfo, we should use http://www.php.net/manual/en/function.finfo-file.php .